### ISSUE TITLE
DefaultScalaModule causing ArrayIndexOutOfBoundsException when object has a method returning a lambda

### ISSUE THREAD

**Issue by jvelliott** [FasterXML/jackson-module-scala#221]
Using the following dependencies on JDK 1.8.0_60

```xml
<dependencies>
        <dependency>
            <groupId>junit</groupId>
            <artifactId>junit</artifactId>
            <version>4.12</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-core</artifactId>
            <version>2.6.1</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>2.6.1</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-annotations</artifactId>
            <version>2.6.1</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.module</groupId>
            <artifactId>jackson-module-scala_2.11</artifactId>
            <version>2.6.1</version>
        </dependency>
</dependencies>
```

Reproducible with the following small testcase:
```java
package test;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.scala.DefaultScalaModule;
import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.function.Function;

/**
 * @author jvelliott
 * @since 8/27/15
 */
@RunWith(JUnit4.class)
public class ObjectMapperFail {
    public class ClassWithALambda {
        public String yay = "yay";

        public Function<String, String> boo() {
            return (input) -> "=(((" + input + ")))=";
        }
    }

    public class ClassNoLambda {
        public String yay = "yay";

        public Function<String, String> boo() {
            return new Function<String, String>() {
                public String apply(String input) {
                    return "=(((" + input + ")))=";
                }
            };
        }
    }

    @Test
    public void testLambdaBroken() throws Exception {
        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new DefaultScalaModule());

        ClassWithALambda cl = new ClassWithALambda();

        String val = mapper.writeValueAsString(cl);  //Fails with ArrayIndexOutOfBoundsException

        Assert.assertNotNull(val);
        Assert.assertEquals("{\"yay\":\"yay\"}", val);
    }

    @Test
    public void testNoLambdaIsOK() throws Exception {
        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new DefaultScalaModule());

        ClassNoLambda cnl = new ClassNoLambda();

        String val = mapper.writeValueAsString(cnl);

        Assert.assertNotNull(val);
        Assert.assertEquals("{\"yay\":\"yay\"}", val); //Passes!
    }
}
```

testLambdaBroken() produces the following exception, and fails:
```
java.lang.ArrayIndexOutOfBoundsException: 28270
	at com.thoughtworks.paranamer.BytecodeReadingParanamer$ClassReader.accept(BytecodeReadingParanamer.java:563)
	at com.thoughtworks.paranamer.BytecodeReadingParanamer$ClassReader.access$200(BytecodeReadingParanamer.java:338)
	at com.thoughtworks.paranamer.BytecodeReadingParanamer.lookupParameterNames(BytecodeReadingParanamer.java:103)
	at com.thoughtworks.paranamer.CachingParanamer.lookupParameterNames(CachingParanamer.java:90)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$.com$fasterxml$jackson$module$scala$introspect$BeanIntrospector$$getCtorParams(BeanIntrospector.scala:44)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$$anonfun$1.apply(BeanIntrospector.scala:58)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$$anonfun$1.apply(BeanIntrospector.scala:58)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:252)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:252)
	at scala.collection.Iterator$class.foreach(Iterator.scala:742)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1194)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:252)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$.com$fasterxml$jackson$module$scala$introspect$BeanIntrospector$$findConstructorParam$1(BeanIntrospector.scala:58)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$$anonfun$2$$anonfun$apply$5.apply(BeanIntrospector.scala:176)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$$anonfun$2$$anonfun$apply$5.apply(BeanIntrospector.scala:170)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$$anonfun$2.apply(BeanIntrospector.scala:170)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$$anonfun$2.apply(BeanIntrospector.scala:169)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:252)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:252)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:252)
	at scala.collection.immutable.List.flatMap(List.scala:344)
	at com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$.apply(BeanIntrospector.scala:169)
	at com.fasterxml.jackson.module.scala.introspect.ScalaAnnotationIntrospector$._descriptorFor(ScalaAnnotationIntrospectorModule.scala:22)
	at com.fasterxml.jackson.module.scala.introspect.ScalaAnnotationIntrospector$.fieldName(ScalaAnnotationIntrospectorModule.scala:30)
	at com.fasterxml.jackson.module.scala.introspect.ScalaAnnotationIntrospector$.findImplicitPropertyName(ScalaAnnotationIntrospectorModule.scala:78)
	at com.fasterxml.jackson.databind.introspect.AnnotationIntrospectorPair.findImplicitPropertyName(AnnotationIntrospectorPair.java:467)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector._addFields(POJOPropertiesCollector.java:351)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.collectAll(POJOPropertiesCollector.java:283)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.getJsonValueMethod(POJOPropertiesCollector.java:169)
	at com.fasterxml.jackson.databind.introspect.BasicBeanDescription.findJsonValueMethod(BasicBeanDescription.java:222)
	at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findSerializerByAnnotations(BasicSerializerFactory.java:355)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:210)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:153)
	at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1199)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1153)
	at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:481)
	at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:679)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:107)
	at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:3525)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:2915)
	at test.ObjectMapperFail.testLambdaBroken(ObjectMapperFail.java:45)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
```

**Comment by jvelliott** [comment]
FWIW: adding ```@JsonIgnore``` to the method has no affect.

**Comment by christophercurrie** [comment]
This is quite bizarre; the failure seems to be in Paranamer's introspection of constructor parameters, which neither of your test classes have. It's possible paranamer is being used incorrectly, but it's very odd to me that the lambda would be the cause. I'll dig further and see what I can understand.

**Comment by mbknor** [comment]
We're seeing lots of problems related to paranamer when using the DefaultScalaModule.

We see problems e.g when starting Dropwizard 0.8.4 applications.

If DefaultScalaModule is loaded, it fails starting up while parsing yaml-config.

It crashes when it analyzes this class:
https://github.com/dropwizard/dropwizard/blob/master/dropwizard-util/src/main/java/io/dropwizard/util/Duration.java#L72

which has this method:
```java
@JsonCreator public static Duration parse(String duration)
```

We also see other issues, like ArrayIndexOutOfBoundsException, when serializing pure java-POJOs.

We tried to create a copy of of DefaultScalaModule using a modified ScalaAnnotationIntrospectorModule with disabled ParanamerAnnotationIntrospector and everything works fine (dropwizard config, java pojos and scala case classes).

So my question is? Why is paranamer used by the scala module?

here are the modified classes:

```scala

/**
 * Copy of DefaultScalaModule with disabled Paranamer - since it causes lots of problems
 */
class FixedDefaultScalaModule extends JacksonModule
  with IteratorModule
  with EnumerationModule
  with OptionModule
  with SeqModule
  with IterableModule
  with TupleModule
  with MapModule
  with SetModule
  with ScalaNumberDeserializersModule
  with FixedScalaAnnotationIntrospectorModule
  with UntypedObjectDeserializerModule
  with EitherModule
  {
    override def getModuleName = "FixedDefaultScalaModule"
}

/**
 * Copy of ScalaAnnotationIntrospectorModule with paranamer disabled
 */
trait FixedScalaAnnotationIntrospectorModule extends JacksonModule {
  //this += { _.appendAnnotationIntrospector(new ParanamerAnnotationIntrospector()) }
  this += { _.appendAnnotationIntrospector(ScalaAnnotationIntrospector) }
}
```

**Comment by jarreds** [comment]
Seeing the same issue on DW 0.9.0-rc4. Confirmed that fix above works there too.

**Comment by mbknor** [comment]
I have created new specific ticket #229 for the static @JsonCreator method issue

**Comment by nbauernfeind** [comment]
I believe that this is now fixed in 2.6.5 and 2.7.2. @jvelliott Any chance you can confirm?

**Comment by jimm-porch** [comment]
I was able to reproduce this issue with 2.6.3, 2.6.4, and 2.6.5

**Comment by jimm-porch** [comment]
I can confirm that removing a custom deserializer for `Duration` fixes this in 2.6.3.

**Comment by jimm-porch** [comment]
Other custom deserializers (like `Date` for example) don't trigger the bug, BTW.

**Comment by jvelliott** [comment]
2.6.5 still fails with the same error

2.7.2 appears to have a new exception, now for both methods. Looks like maybe I need to update jackson for 2.7.2 ?

```
java.lang.NoSuchMethodError: com.fasterxml.jackson.databind.JsonMappingException.<init>(Ljava/io/Closeable;Ljava/lang/String;)V

	at com.fasterxml.jackson.module.scala.JacksonModule$class.setupModule(JacksonModule.scala:56)
	at com.fasterxml.jackson.module.scala.DefaultScalaModule.setupModule(DefaultScalaModule.scala:19)
	at com.fasterxml.jackson.databind.ObjectMapper.registerModule(ObjectMapper.java:709)
	at test.ObjectMapperFail.testLambdaBroken(ObjectMapperFail.java:41)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)


java.lang.NoSuchMethodError: com.fasterxml.jackson.databind.JsonMappingException.<init>(Ljava/io/Closeable;Ljava/lang/String;)V

	at com.fasterxml.jackson.module.scala.JacksonModule$class.setupModule(JacksonModule.scala:56)
	at com.fasterxml.jackson.module.scala.DefaultScalaModule.setupModule(DefaultScalaModule.scala:19)
	at com.fasterxml.jackson.databind.ObjectMapper.registerModule(ObjectMapper.java:709)
	at test.ObjectMapperFail.testNoLambdaIsOK(ObjectMapperFail.java:54)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
```

**Comment by cowtowncoder** [comment]
@jvelliott Minor versions of Jackson components do need to match, in general. Some combinations may work, but it is safer to align minor versions.

**Comment by jvelliott** [comment]
I still see the same exception when running with 

```
      <dependency>
            <groupId>com.fasterxml.jackson.module</groupId>
            <artifactId>jackson-module-scala_2.10</artifactId>
            <version>2.7.2</version>
        </dependency>
```


I've also tried running with (note the artifactId)
```
      <dependency>
            <groupId>com.fasterxml.jackson.module</groupId>
            <artifactId>jackson-module-scala_2.11</artifactId>
            <version>2.7.2</version>
        </dependency>
```

but I get the following exception. 

```
java.lang.NoSuchMethodError: scala.util.matching.Regex.unapplySeq(Ljava/lang/CharSequence;)Lscala/Option;
	at com.fasterxml.jackson.module.scala.JacksonModule$.version$lzycompute(JacksonModule.scala:30)
	at com.fasterxml.jackson.module.scala.JacksonModule$.version(JacksonModule.scala:26)
	at com.fasterxml.jackson.module.scala.JacksonModule$class.version(JacksonModule.scala:49)
	at com.fasterxml.jackson.module.scala.DefaultScalaModule.version(DefaultScalaModule.scala:19)
	at com.fasterxml.jackson.databind.ObjectMapper.registerModule(ObjectMapper.java:714)
	at test.ObjectMapperFail.testLambdaBroken(ObjectMapperFail.java:111)
```

Perhaps some other conflicting dependency?

**Comment by nbauernfeind** [comment]
What version of scala are you using? You cannot mix and match scala versions. Based on your exception when using jackson-module-scala_2.11 I think you're using scala 2.10. You cannot mix and match scala versions. I too see that Paranamer is causing a few issues, and is likely related to this issue.

**Comment by nbauernfeind** [comment]
I can confirm that issue that the original poster provided is fixed as of 2.7.2, it is broken for 2.6.x however (and will probably remain that way). There are other outstanding issues related to Paranamer (reported in #229). @jvelliott you need to make sure that you're using Jackson 2.7.2 for all of your com.fasterxml.jackson.* dependencies (not just the scala module). 

If there are any lingering related issues (though not related to the Paranamer issue) please open a new issue with instructions on how to reproduce.
